### ISSUE TITLE
RedisDynoQueue.pop() not returning HTTP tasks that are in the queue

### ISSUE THREAD

**Issue by rickfish** [Netflix/dyno-queues#31]
When we use the Redis implementation (not MySql), we have HTTP tasks that are in SCHEDULED status but the pop() method of RedisDynoQueue is not returning them. The reason these SCHEDULED tasks are not popped is because the RedisMock class's zAdd method for these tasks for the HTTP UNACK queue returns 0. That causes the task to not be popped in the _pop method of RedisDynoQueue. These tasks become stranded and will never be returned on a poll.

 Seems to be a bug with RedisSortedSetCache. It is trying to maintain two maps, the cache map and the scores map. The scores map value is a Map and the cache map value is a Set. When tasks are added to the UNACK.HTTP cache after  they have already been added before, a duplicate entry is added to the cache map's Set but the scores entry is  just overwritten (since the put does not allow dups). Then when the entry is removed, only one is removed from the cache, leaving one (or more) and the scores entry is deleted entirely.
 
 The Comparator of the cache map is the culprit. It does comparisons using the scores map, not the cache map, so the contract of the SortedSet that says it won't allow duplicates is broken, at least for the way that Conductor is using that class. The contains() method uses the Comparator to determine if the entry already exists and because of the Comparator the contains() will sometimes return false even if there is an entry. Conductor uses this contains() method.
 
 When the pop() method of DynoQueueDAO is called, it calls the pop() method of the “conductor_queues.test.UNACK.HTTP.c” RedisDynoQueue which calls its internal _pop() method. In its _pop() method, it does this for each task in the UNACK.HTTP queue:

                 long added = quorumConn.zadd(unackQueueName, unackScore, msgId, zParams);
                if(added == 0){
                       monitor.misses.increment();
                       continue;
                }

So if ‘added=0’, that task will not be added to the tasks to pop. In the zadd method of RedisMock, the contains() method of the cache map is used to see if the task exists in the cache. If it does, it returns 0. The contains() method uses the Comparator specified by RedisSortedSet.set() method which doesn’t just check the task id of the others in the set, it also will return a 0 (equals) if both scores for those tasks are null or if the score being added is greater than any other in the set, which they are at that point. So it returns 0 because either both scores are null or the to-be-added score is greater than the highest in the set and it is not popped. 

Once there are duplicates, that causes issues that keep SCHEDULED tasks from being popped. 
 
 I am not sure where to go with this. Should Conductor use a different Redis class for its persistence of the tasks in the queue or is this something that should be changed in RedisMock?

**Comment by rickfish** [comment]
Sorry, after looking where I posted this issue, I realized that I should have mentioned that the issue happens when the dyno-queues are being used in Conductor.

**Comment by gansa1986** [comment]
we have configured the redis DB to the netflix docker containers. And a simple workflow with HTTP task is struck in scheduled status for long time. Any ideas?

**Comment by james-deee** [comment]
@rickfish why was this closed out? We're seeing tasks being stuck in a SCHEDULED state

**Comment by rickfish** [comment]
I believe I closed this because I was asked to open it on the Redis issues list. I honestly don't remember but nothing was ever resolved. Once we started using the SQL database vs memory we didn't need to pursue further.

**Comment by christianscheuer** [comment]
Did anybody find out how to solve this?
We're running Conductor on GKE with a setup like this docker-compose (albeit on Kubernetes pods/services):
https://github.com/flaviostutz/conductor-server

In the Conductor UI we can clearly see that SIMPLE tasks get scheduled, and they appear in the "Task Queue" - however when polling from a worker (like this:) it always returns <b>"204 No Content"</b> even if there are items in the queue so the tasks never get run:
```
curl -X GET http://conductor-server:8080/api/tasks/poll/task_name_here
```

@rickfish would you be able to give us some pointers on how you got this fixed? Ie. what would be the way to configure Conductor/Dynomite/dyno-queues to run on SQL instead (if this is the easiest way to fix the issue)

**Comment by rickfish** [comment]
Right, we never got this addressed as it is a problem with the redis dyno-queues, not conductor itself if I remember correctly. We instead decided that for anything that has to work well we would use MySQL (and since then we are now using Postgres).

Our setup for Postgres is this:
```
db=postgres
jdbc.url=jdbc:postgresql://<host>:<port>/<dbname>?ssl=false&currentSchema=<schema-name>
jdbc.username=<username>
jdbc.password=<password>
```

**Comment by christianscheuer** [comment]
Thanks @rickfish. I actually tried switching to MySQL as a provider but it was a nightmare due to migrations in Conductor's source code seemingly being incompatible with all the mysql distros I could find.
I got the polling to work when upgrading to newest version of Conductor (v2.25.5), although it has other issues (they seem to be transitioning away from the ack endpoint, but right now it means that more than one worker can receive the same task for processing). This was using dyno-queues with v2.25.5.
It's good to know that Postgres can be used - I'll try that out if this isn't fixed:
https://github.com/Netflix/conductor/pull/1534#issuecomment-596170693
